Feature/reindex crawler for files

jareddreyerss opened the pull request:
# SwiftypeFileCrawlerExtension

- added new File crawler to handle indexing/reindexing of Files
- set config to whitelist `File::class` filetypes, this allows developers to control which files they want to index.
- added unit tests to cover the File extension and whitelisting.
- Updated read me docs.

jareddreyerss committed:
Creation of new File Crawler extension for Files

jareddreyerss committed:
Added method to check if file being indexed is in thewhitelist

jareddreyerss committed:
Set up new fake service and tests for FileIndexCrawlerExtentsion

jareddreyerss committed:
Added new tests to cover whitelisted file config

Included in the updated fixtures is 1 PDF and 1 Image (jpg file) and their accompanying physical. This is because SilverStripe asset store cannot publish non-existent files, which are needed for our tests.

These fixtures allow us to test reindexing of 'whitelisted' file types.

Also tidy ups on code comments was also applied in this commit (where no context or incorrect ordering of comments are used).

jareddreyerss committed:
Set fake SwiftypeFile to use a default config for whitelist

fake SwiftypeFile::class now sets whitelist to 'pdf' this saves us specifying in our unit tests which we only use 'pdf' and 'jpg' to test.

jareddreyerss committed:
Reset all on hooks to use versioned hooks

Reset all onBefore and onAfter hooks e.g. from `onbeforeWrite` to `onBeforePublish()`, as didn't want anything that wasn't published in our index.

Added a helper check to invalidate the collatedURLs array if a file filetype is not in our whitelist.

Tried to tidy up all references to 'page' in FileCrawler extension but may have missed some xD.

jareddreyerss committed:
Added instructions to setup SwiftypeFileCrawlerExtension

jareddreyerss committed:
Clearing cache after publish

`FlushCache()` does not exist so was not calling the correct clearCache method, so changed all references to use `clearCacheAll()`.

Commit:
FIX tests on when to clear crawl queue

chrispenny reviewed (changes requested):
Thanks again for the PR! Adding support for crawling Files sounds like a great idea.

**Feeback:**

The contents of `SwiftypeFileCrawlerExtension` are almost identical to `SwiftypeSiteTreeCrawlerExtension`. It looks like the unique part is `collateUrls()`, but the difference there is really only one extra `if` statement, and that we need to grab the `Link` from a different method.

To me, this would indicate that we need a shared class that we can then extend for other purposes.

Something like:
- `SwiftypeCrawlerExtension`: Which contains all of the `DataObject` agnostic functionality. Think about this class as being one that you could apply to any type of `DataObject`.
- `SwiftypeSiteTreeCrawlerExtension`: Which extends `SwiftypeCrawlerExtension` and updates only what it needs to in order to make it work with a `SiteTree` object.
- `SwiftypeFileCrawlerExtension`: Which extends `SwiftypeCrawlerExtension` and updates only what it needs to in order to make it work with a `File` object.

Instead of calling your extra method `checkFileIsToBeReindexed()`, you could instead create a more generic method called (something like) `recordCanBeIndexed()`. `SwiftypeCrawlerExtension` can then implement this method by simply returning `true`. `SwiftypeSiteTreeCrawlerExtension` and `SwiftypeFileCrawlerExtension` can then implement the method and perform additional checks as needed.

Similarly, within `collateUrls()`, instead of us just calling:
```
$link = $owner->getAbsoluteLiveLink(false);
```
You could call another generic method in your extension. The generic version might then look for the existence of `Link` methods on the `$owner` and return the most appropriate result, or throw and Exception if no appropriate method exists.

chrispenny commented:
Is this feature still desired?

jareddreyerss commented:
> Is this feature still desired?

its still in use on Westpac afaik I just never got a chance to look at this 100% to complete the feedback.
So I guess it is a question of do we see value merging this back into the module or we happy to stick with the fork.

cc @satrun77

satrun77 commented:
Yes, we need it.

chrispenny commented:
"Yes, we need it" and the team plan to put some time aside soon to action the feedback? Or "Yes, we need it" but don't have capacity to action the feedback?

I could pick it up for you, but I'll probably target this change to Silverstripe 5 just to keep it simple for myself. Does that sound ok @jareddreyerss @satrun77?

satrun77 commented:
Yes I will put time to action the feedback after you have done cms 5 work

Cambis committed:
Refactor swiftype extensions into generic instance

- The extensions for File and SiteTree had the essential the same functionality with one or two lines of difference. The main functionality was pulled into new class SwiftTypeCrawlerExtension
- New class provides new methods to get object links and exclude records from indexing

Cambis committed:
Update owner link method

- Links should come from the live/published table only. Update the owner link method to explicitally search the live table, adapted from SiteTree::getAbsoluteLiveLink()
- Minor cleaning of file crawler tests

Cambis commented:
@chrispenny ready for review :)

chrispenny commented on src/Extensions/SwiftTypeCrawlerExtension.php:
```diff
@@ -0,0 +1,313 @@
+/**
+ * @property DataObject|$this $owner
+ */
+class SwiftTypeCrawlerExtension extends DataExtension
```
`Swiftype` rather than `SwiftType`.

I also don't think we want anyone applying this extension to their classes? If that's the case, then this should be `abstract`. This forces folks to pick either the `File` or `SiteTree` extensions specifically.

Cambis replied:
Tbh this has been my main problem with swiftype over the years, can't seem to spell it correctly!

chrispenny commented on src/Extensions/SwiftTypeCrawlerExtension.php:
```diff
@@ -0,0 +1,313 @@
+     *
+     * @throws \InvalidArgumentException if owner does not implement {@link AbsoluteLink()}
+     */
+    protected function getOwnerLink(): ?string
+    {
+        if (!method_exists($this->getOwner(), 'AbsoluteLink')) {
+            throw new \InvalidArgumentException('Invalid object passed, object must implement AbsoluteLink()');
+        }
+
+        $live = Versioned::get_by_stage(get_class($this->getOwner()), Versioned::LIVE)->byID($this->getOwner()->ID);
+
+        if ($live) {
+            $link = $live->AbsoluteLink();
+        } else {
+            $link = null;
+        }
+
+        return $link;
+    }
```
I think an `abstract` method would be more appropriate here. The idea of this class should now be that you really have no idea what sort of Model this class will be extended to serve, so any code that isn't "one size fits all" should belong to the classes that extend this `abstract`.

This code can then be moved into `SwiftypeFileCrawlerExtension`, and you've already covered the `SwiftypeSiteTreeCrawlerExtension`.

chrispenny commented on src/Extensions/SwiftypeFileCrawlerExtension.php:
```diff
@@ -0,0 +1,23 @@
+        // only reindex file types we need.
+        $fileType = File::get_file_extension($this->getOwner()->Filename);
+
+        return in_array($fileType, $this->getOwner()->config()->get('reindex_files_whitelist'), true);
```
Not to be overly woke, but I'd prefer the use of "allow" over "white". It's kind of an easy thing to change to be [doing our part]. This comment would go for the `README.md`, tests, etc.

This doesn't have to be added now (it can be an improvement later), but it would also be good to have support for file categories ("images", "documents", etc).

Cambis replied:
I definitely prefer allowlist as well, will change

chrispenny commented on src/Extensions/SwiftypeFileCrawlerExtension.php:
```diff
@@ -0,0 +1,23 @@
+
+use SilverStripe\Assets\File;
+
+/**
+ * @property File|$this $owner
+ */
```
Don't need this, as it's defined in `SwiftTypeCrawlerExtension`.
```suggestion
```

You're using `getOwner()` in this file though, and we know this extension is meant to be applied to `File`, so it might be nice to add an `@method File|$this getOwner()`

chrispenny commented on src/Extensions/SwiftypeSiteTreeCrawlerExtension.php:
```diff
@@ -2,283 +2,15 @@
-use SilverStripe\Core\Config\Config;
-use SilverStripe\Versioned\Versioned;
 
 /**
  * @property SiteTree|$this $owner
  */
```
Don't need this, as it's defined in `SwiftTypeCrawlerExtension`.
```suggestion
```

You're using `getOwner()` in this file though, and we know this extension is meant to be applied to `SiteTree`, so it might be nice to add an `@method SiteTree|$this getOwner()`.

chrispenny commented on src/Extensions/SwiftTypeCrawlerExtension.php:
```diff
@@ -0,0 +1,313 @@
+    /**
+     * Extend this method in order to check if a record can be indexed. Defaults to true.
+     */
+    protected function recordCanBeIndexed(): bool
+    {
+        return true;
+    }
```
I'm 50/50 on this one being `abstract` or not. This code technically can work for any type of Model that this class gets extended to serve. I think the question is more "do we want developers to have to think about whether or not a model can be indexed?".

I'll leave the choice to you. I really am on the fence.

chrispenny reviewed (changes requested):
Awesome effort on the refactor, @Cambis! Things are looking great!

I haven't had a chance to perform any functional testing yet, but I have some minor initial feedback.

Cambis committed:
Further abstraction on crawler classes

Cambis commented:
Thanks for the feedback back at you @chrispenny

chrispenny reviewed (approved):
Apologies it took me so long to get back around for this, but it all looks amazing :D Thank you for your efforts, team!

I'll also try to get this merged up for SS 5 as well.

chrispenny commented:
Tagged: https://github.com/silverstripe-terraformers/silverstripe-swiftype/releases/tag/1.2.0
